// transaction-sms-parser/dist/lib.js
var i = /* @__PURE__ */ ((t) => (t.CARD = "CARD", t.WALLET = "WALLET", t.ACCOUNT = "ACCOUNT", t))(i || {}), u = /* @__PURE__ */ ((t) => (t.AVAILABLE = "AVAILABLE", t.OUTSTANDING = "OUTSTANDING", t))(u || {});
const A = [
  "avbl bal",
  "available balance",
  "available limit",
  "available credit limit",
  "avbl. credit limit",
  "limit available",
  "a/c bal",
  "ac bal",
  "available bal",
  "avl bal",
  "updated balance",
  "total balance",
  "new balance",
  "bal",
  "avl lmt",
  "available"
], y = ["outstanding"], L = ["paytm", "simpl", "lazypay", "amazon_pay"], N = ["upi", "ref no", "upi ref", "upi ref no", "ref"], m = [
  {
    regex: /credit\scard/g,
    word: "c_card",
    type: i.CARD
  },
  {
    regex: /amazon\spay/g,
    word: "amazon_pay",
    type: i.WALLET
  },
  {
    regex: /uni\scard/g,
    word: "uni_card",
    type: i.CARD
  },
  {
    regex: /niyo\scard/g,
    word: "niyo",
    type: i.ACCOUNT
  },
  {
    regex: /slice\s*(credit|credt)?\s*card/g,
    word: "slice_card",
    type: i.CARD
  },
  {
    regex: /one\s*(credit|credt)?\s*card/g,
    word: "one_card",
    type: i.CARD
  }
], T = [
  "@BARODAMPAY",
  "@rbl",
  "@idbi",
  "@upi",
  "@aubank",
  "@axisbank",
  "@bandhan",
  "@dlb",
  "@indus",
  "@kbl",
  "@federal",
  "@sbi",
  "@uco",
  "@citi",
  "@citigold",
  "@dlb",
  "@dbs",
  "@freecharge",
  "@okhdfcbank",
  "@okaxis",
  "@oksbi",
  "@okicici",
  "@yesg",
  "@hsbc",
  "@idbi",
  "@icici",
  "@indianbank",
  "@allbank",
  "@kotak",
  "@ikwik",
  "@unionbankofindia",
  "@uboi",
  "@unionbank",
  "@paytm",
  "@ybl",
  "@axl",
  "@ibl",
  "@sib",
  "@yespay"
], v = (t) => !Number.isNaN(Number(t)), R = (t) => {
  const [e, c] = [t[0], t[t.length - 1]];
  let n = Number.isNaN(Number(c)) ? t.slice(0, -1) : t;
  return n = Number.isNaN(Number(e)) ? n.slice(1) : n, n;
}, I = (t) => {
  const e = t.replace("ac", "");
  return Number.isNaN(Number(e)) ? "" : e;
}, h = (t) => {
  let e = t.toLowerCase();
  e = e.replace(/!/g, ""), e = e.replace(/:/g, " "), e = e.replace(/\//g, ""), e = e.replace(/=/g, " "), e = e.replace(/[{}]/g, " "), e = e.replace(/\n/g, " "), e = e.replace(/\r/g, " "), e = e.replace(/ending /g, ""), e = e.replace(/x|[*]/g, ""), e = e.replace(/is /g, ""), e = e.replace(/with /g, ""), e = e.replace(/no. /g, ""), e = e.replace(/\bac\b|\bacct\b|\baccount\b/g, "ac"), e = e.replace(/rs(?=\w)/g, "rs. "), e = e.replace(/rs /g, "rs. "), e = e.replace(/inr(?=\w)/g, "rs. "), e = e.replace(/inr /g, "rs. "), e = e.replace(/rs. /g, "rs."), e = e.replace(/rs.(?=\w)/g, "rs. "), e = e.replace(/debited/g, " debited "), e = e.replace(/credited/g, " credited ");
  for (const c of m)
    e = e.replace(c.regex, c.word);
  return e.split(" ").filter((c) => c !== "");
}, g = (t) => {
  let e = [];
  return typeof t == "string" ? e = h(t) : e = t, e;
}, p = (t) => {
  const [e, c] = t.split(".");
  return `${e}.${(c ?? "").padEnd(2, "0")}`;
}, O = (t, e, c = 1) => {
  const a = t.split(e, 2)[1];
  if (a) {
    const r = /[^0-9a-zA-Z]+/gi;
    return a.trim().split(r, c).join(" ");
  }
  return "";
}, B = (t) => {
  let e = "";
  const c = t.findIndex(
    (a) => a === "card" || m.filter((r) => r.type === i.CARD).some((r) => r.word === a ? (e = r.word, !0) : !1)
  ), n = { type: null, name: null, number: null };
  return c !== -1 ? (n.number = t[c + 1], n.type = i.CARD, Number.isNaN(Number(n.number)) ? {
    type: e ? n.type : null,
    name: e,
    number: null
  } : n) : { type: null, name: null, number: null };
}, x = (t) => {
  const e = g(t);
  let c = -1, n = {
    type: null,
    name: null,
    number: null
  };
  for (const [a, r] of e.entries())
    if (r === "ac")
      if (a + 1 < e.length) {
        const s = R(
          e[a + 1]
        );
        if (Number.isNaN(Number(s)))
          continue;
        c = a, n.type = i.ACCOUNT, n.number = s;
        break;
      } else
        continue;
    else if (r.includes("ac")) {
      const s = I(r);
      if (s === "")
        continue;
      c = a, n.type = i.ACCOUNT, n.number = s;
      break;
    }
  if (c === -1 && (n = B(e)), !n.type) {
    const a = e.find((r) => L.includes(r));
    a && (n.type = i.WALLET, n.name = a);
  }
  if (!n.type) {
    const a = m.filter((r) => r.type === i.ACCOUNT).find((r) => e.includes(r.word));
    n.type = (a == null ? void 0 : a.type) ?? null, n.name = (a == null ? void 0 : a.word) ?? null;
  }
  return n.number && n.number.length > 4 && (n.number = n.number.slice(-4)), n;
}, E = (t, e, c) => {
  let n = "", a = !1, r = 0, s = "", l = t;
  for (; l < c; ) {
    if (s = e[l], s >= "0" && s <= "9")
      a = !0, n += s;
    else if (a) {
      if (s === ".") {
        if (r === 1)
          break;
        n += s, r += 1;
      } else if (s !== ",")
        break;
    }
    l += 1;
  }
  return n;
}, D = (t, e = u.AVAILABLE) => {
  const n = `(${(e === u.AVAILABLE ? A : y).join("|")})`.replace("/", "\\/"), a = "([\\d]+\\.[\\d]+|[\\d]+)";
  let r = new RegExp(`${n}\\s*${a}`, "gi"), s = t.match(r);
  if (s && s.length > 0) {
    const l = s[0].split(" ").pop();
    return Number.isNaN(Number(l)) ? "" : l;
  }
  if (r = new RegExp(`${a}\\s*${n}`, "gi"), s = t.match(r), s && s.length > 0) {
    const l = s[0].split(" ")[0];
    return Number.isNaN(Number(l)) ? "" : l;
  }
  return null;
}, f = (t, e = u.AVAILABLE) => {
  const n = g(t).join(" ");
  let a = -1, r = "";
  const s = e === u.AVAILABLE ? A : y;
  for (const b of s)
    if (a = n.indexOf(b), a !== -1) {
      a += b.length;
      break;
    }
  let l = a, o = -1, d = n.substr(l, 3);
  for (l += 3; l < n.length; ) {
    if (d = d.slice(1), d += n[l], d === "rs.") {
      o = l + 2;
      break;
    }
    l += 1;
  }
  return o === -1 ? (r = D(n) ?? "", r ? p(r) : null) : (r = E(o, n, n.length), r ? p(r) : null);
}, w = (t) => {
  const e = g(t), c = e.join(" "), n = {
    merchant: null,
    referenceNo: null
  };
  if (e.includes("vpa")) {
    const r = e.indexOf("vpa");
    if (r < e.length - 1) {
      const s = e[r + 1], [l] = s.replaceAll(/\(|\)/gi, " ").split(" ");
      n.merchant = l;
    }
  }
  let a = "";
  for (let r = 0; r < N.length; r += 1) {
    const s = N[r];
    c.indexOf(s) > 0 && (a = s);
  }
  if (a) {
    const r = O(c, a);
    if (v(r))
      n.referenceNo = r;
    else if (n.merchant) {
      const [s] = r.split(/[^0-9]/gi).sort((l, o) => o.length - l.length)[0];
      s && (n.referenceNo = s);
    } else
      n.merchant = r;
    if (!n.merchant) {
      const s = new RegExp(
        `[a-zA-Z0-9_-]+(${T.join("|")})`,
        "gi"
      ), l = c.match(s);
      if (l && l.length > 0) {
        const o = l[0].split(" ").pop();
        n.merchant = o ?? null;
      }
    }
  }
  return n;
}, C = (t) => {
  const c = g(t).indexOf("rs.");
  if (c === -1)
    return "";
  let n = t[c + 1];
  return n = n.replace(/,/g, ""), Number.isNaN(Number(n)) ? (n = t[c + 2], n = n == null ? void 0 : n.replace(/,/g, ""), Number.isNaN(Number(n)) ? "" : p(n)) : p(n);
}, k = (t) => {
  const e = /(?:credited|credit|deposited|added|received|refund|repayment)/gi, c = /(?:debited|debit|deducted)/gi, n = /(?:payment|spent|paid|used\s+at|charged|transaction\son|transaction\sfee|tran|booked|purchased|sent\s+to|purchase\s+of|spent\s+on)/gi, a = typeof t != "string" ? t.join(" ") : t;
  return c.test(a) || n.test(a) ? "debit" : e.test(a) ? "credit" : null;
}, M = (t) => {
  if (!t || typeof t != "string")
    return {
      account: {
        type: null,
        number: null,
        name: null
      },
      balance: null,
      transaction: {
        type: null,
        amount: null,
        merchant: null,
        referenceNo: null
      }
    };
  const e = h(t), c = x(e), n = f(
    e,
    u.AVAILABLE
  ), a = C(e), s = [n, a, c.number].filter(
    (b) => b !== ""
  ).length >= 2 ? k(e) : null, l = { available: n, outstanding: null };
  c && c.type === i.CARD && (l.outstanding = f(
    e,
    u.OUTSTANDING
  ));
  const { merchant: o, referenceNo: d } = w(t);
  return {
    account: c,
    balance: l,
    transaction: {
      type: s,
      amount: a,
      merchant: o,
      referenceNo: d
    }
  };
}, S = x, $ = f, V = w, U = M, K = C, _ = k;
export {
  i as IAccountType,
  u as IBalanceKeyWordsType,
  S as getAccountInfo,
  $ as getBalanceInfo,
  V as getMerchantInfo,
  K as getTransactionAmount,
  U as getTransactionInfo,
  _ as getTransactionType
};
